File: Structural Patterns/Bridge/Bridge01.js
// Interface de Cores
class Color {
    constructor(color){
        this.color = color;
    }
    getColor(){
        return this.color;
    }
}

// Implementação Cores
class RedColor extends Color{
    constructor(){
        super("vermelho");
    }
}
class GreenColor extends Color{
    constructor(){
        super("verde");
    }
}
class BlueColor extends Color{
    constructor(){
        super("azul");
    }
}
class WhiteColor extends Color{
    constructor(){
        super("branco");
    }
}

// Implementação Forma
class Shape{
    constructor(color){
        this.color = color;
    }

    setColor(color){
        this.color = color;
    }

    draw(){
        throw new Error("Este método precisa ser implementado pela subclasse")
    }
}

// Abstrações Refinadas
class CircleShape extends Shape{
    draw(){
        console.log(`Desenhando um circulo ${this.color.getColor()}.`);
    }
}
class SquareShape extends Shape{
    draw(){
        console.log(`Desenhando um quadrado ${this.color.getColor()}.`);
    }
}
class TriangleShape extends Shape{
    draw(){
        console.log(`Desenhando um triangulo ${this.color.getColor()}.`);
    }
}

// CLiente

const red = new RedColor();
const green = new GreenColor();
const blue = new BlueColor();
const white = new WhiteColor();

const circle = new CircleShape(green);
const square = new SquareShape(red);
const triangle = new TriangleShape(white);

circle.draw()
square.draw()
triangle.draw()

File: Structural Patterns/Bridge/Bridge01.py
from abc import ABC, abstractmethod

# Interface de Cores

class Color(ABC):
    
    @abstractmethod
    def getColor(self):
        pass
    


# Implementação Cores
class RedColor (Color):
    def getColor(self):
        return "vermelho"
    

class GreenColor (Color):
    def getColor(self):
        return "verde"
    

class BlueColor (Color):
    def getColor(self):
        return "azul"
    

class WhiteColor (Color):
    def getColor(self):
        return "branco"
    


# Implementação Forma
class Shape(ABC):
    def __init__(self, color):
        self.color = color
    

    def setColor(self, color):
        self.color = color
    
    @abstractmethod
    def draw(self):
        pass
    


# Abstrações Refinadas
class CircleShape (Shape):
    def draw(self):
        print(f'Desenhando um circulo {self.color.getColor()}.')
    

class SquareShape (Shape):
    def draw(self):
        print(f'Desenhando um quadrado {self.color.getColor()}.')
    

class TriangleShape (Shape):
    def draw(self):
        print(f'Desenhando um triangulo {self.color.getColor()}.')
    


# CLiente

red = RedColor()
green = GreenColor()
blue = BlueColor()
white = WhiteColor()

circle = CircleShape(green)
square = SquareShape(red)
triangle = TriangleShape(white)

circle.draw()
square.draw()
triangle.draw()